pulsar-broker-common/src/main/java/org/apache/pulsar/bookie/rackawareness/BookieRackAffinityMapping.java

```diff
@@ -116,25 +115,32 @@ public synchronized void setConf(Configuration conf) {
         super.setConf(conf);
         MetadataStore store;
         try {
-            store = createMetadataStore(conf);
-            bookieMappingCache = store.getMetadataCache(BookiesRackConfiguration.class);
-            store.registerListener(this::handleUpdates);
-            racksWithHost = bookieMappingCache.get(BOOKIE_INFO_ROOT_PATH).get()
-                    .orElseGet(BookiesRackConfiguration::new);
-            for (Map<String, BookieInfo> bookieMapping : racksWithHost.values()) {
-                for (String address : bookieMapping.keySet()) {
-                    bookieAddressListLastTime.add(BookieId.parse(address));
-                }
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("BookieRackAffinityMapping init, bookieAddressListLastTime {}",
-                            bookieAddressListLastTime);
-                }
-            }
-            updateRacksWithHost(racksWithHost);
-            watchAvailableBookies();
-        } catch (InterruptedException | ExecutionException | MetadataException e) {
+            store = getMetadataStore(conf);
+        } catch (MetadataException e) {
             throw new RuntimeException(METADATA_STORE_INSTANCE + " failed to init BookieId list");
         }
+
+        bookieMappingCache = store.getMetadataCache(BookiesRackConfiguration.class);
+        store.registerListener(this::handleUpdates);
+        bookieMappingCache.get(BOOKIE_INFO_ROOT_PATH)
```

Comment:
this async processing may cause a race condition and bk-client may create ledgers before it updates bookie with rack info, and that can cause ledgers to be created on the wrong bookies and that will not work for us because we are using bookie affinity policy and we can't afford ledgers created on wrong bookies.

Reply:
valid point. I guess this is hard to solve since the interface is synchronous in Bookkeeper. I guess there are ways to workaround the problem and still make it asynchronous. In that case, the initialization would have to somehow happen before the bk-client is made ready.

Reply:
Ok, makes sense. I think we should still change the sync call in the watcher event. It's unlikely to have a cache miss there, though it's technically possible. 

Without changing BK client constructor to reflect the blocking operation that happens, I think we could change the BK provider class to do the creation in background.

Reply:
@merlimat can you please review #22842 for now as it makes bk-client creation async and avoids blocking call at `org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.asyncOpen` which will prevent the deadlock.

Reply:
I think it does not solve the problem completely because the calling thread is still blocked. 

The BK client creation is moved to a background thread and it's ok, thought the thread waiting for the BK client to be created is still blocked.

eg: 
 1. IO thread is initiating the creation of topic
 2. We trigger creation of BK client
 3. background-thread is creating BK client
 4. IO thread is still blocked waiting for completion.

I'm working on a change to return future from BK client factory that would avoid the problem.

Reply:
merging this for now since it's solves a (small) part of the problem

Reply:
@rdhabalia 

>infact this PR should not be merged.

This is solving a problem your PR was not addressing, the use of blocking call in the notification watcher.